src/javascript/class-scanner.ts

```diff
@@ -335,6 +347,221 @@ interface CustomElementDefinition {
   definition?: ElementDefineCall;
 }
 
+class PrototypeMemberFinder implements Visitor {
+  readonly members = new Map<string,
+    {methods: Map<string, ScannedMethod>, properties: Map<string, ScannedProperty>}>();
+  private readonly _document: JavaScriptDocument;
+
+  constructor(document: JavaScriptDocument) {
+    this._document = document;
+  }
+
+  enterExpressionStatement(node: babel.ExpressionStatement) {
+    if (babel.isAssignmentExpression(node.expression)) {
+      this._createMemberFromAssignment(node);
+    } else if (babel.isMemberExpression(node.expression)) {
+      this._createMemberFromMemberExpression(node);
+    }
+  }
+
+  private _createMemberFromAssignment(node: babel.ExpressionStatement) {
+    if (!babel.isAssignmentExpression(node.expression)) {
+      return;
+    }
+
+    const expr = node.expression;
+
+    if (!babel.isMemberExpression(expr.left) ||
+        !babel.isMemberExpression(expr.left.object) ||
+        !babel.isIdentifier(expr.left.property)) {
+      return;
+    }
+
+    const leftExpr = expr.left.object;
+    const leftProperty = expr.left.property;
+    const cls = getIdentifierName(leftExpr.object);
+    if (!cls || getIdentifierName(leftExpr.property) !== 'prototype') {
+      return;
+    }
+
+    const annotation = getJSDocAnnotationForNode(node);
+
+    if (babel.isFunctionExpression(expr.right)) {
+      const method = this._createMethodFromExpression(leftProperty.name,
+        expr.right, annotation);
+      if (method) {
+        this._addMethodToClass(cls, method);
+      }
+    } else {
+      const prop = this._createPropertyFromExpression(leftProperty.name,
+        expr, annotation);
+      if (prop) {
+        this._addPropertyToClass(cls, prop);
+      }
+    }
+  }
+
+  private _addMethodToClass(cls: string, member: ScannedMethod) {
+    let classMembers = this.members.get(cls);
+    if (!classMembers) {
+      classMembers = {
+        methods: new Map<string, ScannedMethod>(),
+        properties: new Map<string, ScannedProperty>()
+      };
+      this.members.set(cls, classMembers);
+    }
+    classMembers.methods.set(member.name, member);
+  }
+
+  private _addPropertyToClass(cls: string, member: ScannedProperty) {
+    let classMembers = this.members.get(cls);
+    if (!classMembers) {
+      classMembers = {
+        methods: new Map<string, ScannedMethod>(),
+        properties: new Map<string, ScannedProperty>()
+      };
+      this.members.set(cls, classMembers);
+    }
+    classMembers.properties.set(member.name, member);
```

Comment:
@rictic the question about unifying two methods was these two.. they are identical except the last line. so i was trying to detect if its a method or a property.

Reply:
What about a MapWithDefault class to cover the common code? Something like:

```typescript
class MapWithDefault<K, V> extends Map<K, V> {
  constructor(private readonly makeDefault: (key: K) => V) {
    super();
  }

  get(key: K): V {
    if (!this.has(key)) {
      this.set(key, this.makeDefault(key));
    }
    return super.get(key)!;
  }
}
```